json_patch: add first implementation only with patch application

commodo opened the pull request:
Initially I wanted to also do a function that generates the JSON patch from
two JSON documents, but even just applying the JSON patch was a bit of
work, especially when needing to satisfy all the test-cases.

This change defines all the operation in the RFC6902. The addition isn't
too big (for the json_patch_apply() function), as part of the heavy lifting
is also done by JSON pointer logic.

All the ops were tested with the test-cases defined at:
  https://github.com/json-patch/json-patch-tests

RFC6902: https://tools.ietf.org/html/rfc6902

However, in order to properly support JSON patch:
* json_object_array_insert_idx() is needed in the JSON object API
* this changeset also changes [slightly] the behavior of `json_object_set{f}()` ; I figured, it may be still doable; I can revert that behavior change [if needed]; the behavior of `json_object_set{f}()` wasn't clarified initially;
* the `json_pointer_set_with_array_cb()` mechanism should be useful to support various JSON array behavior types

Signed-off-by: Alexandru Ardelean <ardeleanalex@gmail.com>

commodo commented:
I can split this up in smaller PRs to reduce the amount of review effort.
Some changes in this set can be moved independently, if that helps.

hawicz commented on tests/test_json_patch.c:
```diff
@@ -0,0 +1,119 @@
+void test_json_patch_using_file(const char *testdir, const char *filename)
+{
+	char full_filename[PATH_MAX];
+	(void)snprintf(full_filename, sizeof(full_filename), "%s/%s", testdir, filename);
```
You need a #include "snprintf_compat.h" at the top of this file to use snprintf.  (otherwise Windows builds break)

hawicz replied:
done in https://github.com/hawicz/json-c/tree/json_patch

hawicz commented on arraylist.c:
```diff
@@ -125,6 +125,27 @@ int array_list_shrink(struct array_list *arr, size_t empty_slots)
+{
+	size_t move_amount;
+
+	if (idx > arr->length)
```
This should be "if (idx >= arr->length)"

commodo replied:
the code below seems to work fine to cover the `idx == arr->length`;
but it makes sense to do `idx >= arr->length` here

so, ack

hawicz commented on json-c.sym:
```diff
@@ -16,8 +16,10 @@ JSONC_PRIVATE {
     array_list_del_idx;
     array_list_free;
     array_list_new;
+    array_list_insert_idx;
```
Please do not add anything to JSONC_PRIVATE.  Unlike the previously existing symbols, if these new ones are internal-only, they shouldn't be exported at all.

commodo replied:
ack

hawicz commented on json_object.h:
```diff
@@ -605,6 +605,27 @@ JSON_EXPORT int json_object_array_add(struct json_object *obj, struct json_objec
+ * fields to objects in code more compact. If you want to retain a reference
+ * to an added object you must wrap the passed object with json_object_get
+ *
+ * The reference count of a replaced object will be decremented.
```
This func won't replace any object; delete this line.

commodo replied:
ack;
this was silly of me

hawicz commented on json_object_private.h:
```diff
@@ -100,6 +100,25 @@ void _json_c_set_last_err(const char *err_fmt, ...);
 
 extern const char *json_hex_chars;
 
+struct json_pointer_get_result {
```
Let's add a json_pointer_private.h, to keep these a little better contained to just json_patch.c and json_pointer.c

hawicz replied:
done in https://github.com/hawicz/json-c/tree/json_patch

hawicz commented on json_patch.c:
```diff
@@ -0,0 +1,249 @@
+	int rc;
+
+	if (!json_object_object_get_ex(patch_elem, "from", &jfrom)) {
+		errno = EINVAL;
```
Let's avoid (ab)using errno to return additional json-c-specific error information.  Instead, consider calling _json_c_set_last_err() with a more detailed message.

hawicz replied:
Also, it would be worthwhile adding a json patch specific error code param to the json_patch_apply() signature, to ensure it's possible for callers to implement the recommended [error handling](https://tools.ietf.org/html/rfc5789#section-2.2) based on how things actually failed.

hawicz replied:
I'm working on this in my fork

hawicz replied:
I added a `struct json_patch_error` arg to `json_patch_apply()`.

hawicz commented on json_patch.c:
```diff
@@ -0,0 +1,249 @@
+	}
+
+	/* errno should be set inside json_object_deep_copy() */
+	if (json_object_deep_copy(base, res, NULL) < 0)
```
It should be up to the caller to copy the incoming object or not.  I can imagine that in many cases creating a copy will be needless overhead.

hawicz replied:
I'm working on this in my fork

hawicz replied:
I've changed the meaning of the args a bit, there is now a `copy_from` arg and a `base` arg.

hawicz commented on json_patch.h:
```diff
@@ -0,0 +1,46 @@
+ * If the patch is not correctly formatted, an error will
+ * be returned.
+ *
+ * The original `base` object will first be copied, and then
```
As I mentioned on the json_object_deep_copy() call in json_patch.c, let's leave this up to the caller to do or not.

hawicz commented on json_patch.c:
```diff
@@ -0,0 +1,249 @@
+int json_patch_apply(struct json_object *base, struct json_object *patch,
+                     struct json_object **res)
+{
+	size_t i;
```
please don't use single character variable named.  e.g. call this "ii" instead.

hawicz replied:
fixed in my fork

hawicz commented on json_patch.c:
```diff
@@ -0,0 +1,249 @@
+		json_object_object_del(jpres->parent, jpres->id.key);
+		return 0;
+	} else {
+		return json_object_put(jpres->obj);
```
Is this the case where you're removing the root object?  If so, doesn't that mean that *res is now invalid, and should be set to NULL?  Is removing the root even a valid thing to do?

hawicz replied:
Fixed at https://github.com/hawicz/json-c/commit/670bf12f9ee5cef46a116c3350b78cf4a358ec2b

hawicz commented on json_patch.c:
```diff
@@ -0,0 +1,249 @@
+	json_object_get(from.obj);
+
+	if (!move) {
+		array_set_cb = json_object_array_insert_idx_cb;
```
What if from.parent isn't an array?  If you're not supporting a parent that's a json_type_object then you should check the type here, before getting as far as calling any json_object_array* functions on the wrong object type.  (and note the limitation in the json_patch_apply() docs)

hawicz replied:
Never mind, I didn't quite understand how this works.  There is a check inside json_pointer for the type of the object, and only when it _is_ an array is this callback used, so this is fine.

commodo commented:
it will take me a while to re-spin everything; i came back from a short vacation;
would it be ok to do a PR with just the first 3 patches?
i'll probably do a PR with just those 3 in the meantime;

commodo commented:
Changelog v1 -> v2:
* only first 3 commits changed [for now] ; also moved to PR https://github.com/json-c/json-c/pull/704 in case that simplifies things; i'm also fine to close that
* https://github.com/json-c/json-c/compare/a0c7225cd907a1cf634a30584c3bb2a1531170fa..5ff17b986defe4fc0f604eb141a876fdf67e46be
* changed `idx > arr->length` -> `idx >= arr->length`
* removed `array_list_insert_idx` from `JSONC_PRIVATE`
* removed ` The reference count of a replaced object will be decremented.` docstring from `json_object_array_insert_idx()` ; the patch may show up like it's `json_object_array_put_idx()`
* updated test-case for json_object_array_insert_idx() to insert last element as well

hawicz commented:
Fyi, I adjusted some of these changes, and I'm adding additional fixes over at https://github.com/hawicz/json-c/tree/json_patch

commodo commented:
> Fyi, I adjusted some of these changes, and I'm adding additional fixes over at https://github.com/hawicz/json-c/tree/json_patch

Awesome :)
Thanks

Apologies I did not get a chance to do any updates on this.
Hopefully, all the work (I did) helps.
I'm having a bit of a tough time remembering implementation details on this; but I'll try to take a look.

hawicz commented:
Hang tight for a bit, I've got a number of local changes that I'll be pushing to my fork which address my earlier review comments.  (it might take a day or two)

hawicz commented:
@commodo, I believe this is in a pretty good shape to merge now.  Would you care to give it a once over before I do so?

commodo commented on CMakeLists.txt:
```diff
@@ -88,7 +88,8 @@ option(ENABLE_RDRAND                  "Enable RDRAND Hardware RNG Hash Seed."
 option(DISABLE_EXTRA_LIBS             "Avoid linking against extra libraries, such as libbsd." OFF)
-option(DISABLE_JSON_POINTER           "Disable JSON pointer (RFC6901) support."               OFF)
+option(DISABLE_JSON_POINTER           "Disable JSON pointer (RFC6901) and JSON patch support." OFF)
+option(DISABLE_JSON_PATCH             "Disable JSON patch (RFC6902) support."                 OFF)
```
nice: thanks for these :)
i've been also meaning to do this;
i was a bit undecided whether to do them in CMake or classic Make

hawicz replied:
er... actually _you_ did that: 393c98fb  :) 
I was rather surprised github allowed me to push to a branch in your fork.  Did you explicitly allow that?  
Anyway, it allowed for both of our changes to be rolled into this same PR, at the expense of a bit of confusion over the origin of the code, at least if you're just looking at the overall diffs.

commodo replied:
> Did you explicitly allow that?

no, i went with defaults;

commodo commented on json_patch.c:
```diff
@@ -0,0 +1,332 @@
+	rc = __json_patch_apply_remove(&jpres);
+	if (rc < 0)
+		_set_err(EINVAL, "Unable to remove path referenced by 'path' field");
+	// This means we removed and freed the root object, i.e. *res
```
did i add the `freed` typo?

hawicz replied:
No, I wrote that (intentionally past tense)

commodo reviewed (commented):
All looks good to me :)

I'm mostly counting on the json_patch tests to pass.
IIRC: they're taken from some official JSON-patch repo.

commodo commented:
Wooo
Thanks :)

I know this sounds a bit ¯\\\_(ツ)\_/¯  , but with this, I'm now finished with json-c (for a while).

Contextually: 
* json-c is used by OpenWrt (as a core lib)
* I was working for a company a few years ago, and we were using OpenWrt in a project
* We were experimenting with solutions to address issues of large JSON files (50-100 MB JSON files).
  * Someone might ask whether JSON is the best format for large data-sets
  * For us, that was the situation between configuration coming from the cloud and having many network devices getting their configs (with routing info and stuff)
* JSON patch was a solution for this, but we needed something fast
  *  I don't know what they ended up doing (I left at some point)
* I only managed to implement JSON pointer (while still working on that project)
* I never thought I'd finish JSON patch; I started it sometime after I left (since it seemed fun to do)

Thanks again :)
Alex
